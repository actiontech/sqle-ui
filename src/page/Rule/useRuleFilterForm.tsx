import { Select } from 'antd';
import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import { IRuleResV1 } from '../../api/common';
import useDatabaseType from '../../hooks/useDatabaseType';
import useGlobalRuleTemplate from '../../hooks/useGlobalRuleTemplate';
import useProject from '../../hooks/useProject';
import useRuleTemplate from '../../hooks/useRuleTemplate';

export enum RuleUrlParamKey {
  projectName = 'projectName',
  ruleTemplateName = 'ruleTemplateName',
}

const useRuleFilterForm = (
  getProjectTemplateRules: (
    projectName?: string,
    ruleTemplateName?: string
  ) => Promise<IRuleResV1[]>,
  getGlobalTemplateRules: (ruleTemplateName?: string) => Promise<IRuleResV1[]>
) => {
  const { t } = useTranslation();
  const location = useLocation();

  const [dbType, setDbType] = useState<string | undefined>(undefined);
  const { updateDriverNameList, driverNameList, generateDriverSelectOptions } =
    useDatabaseType();
  const { updateProjectList, generateProjectSelectOption } = useProject();
  const { ruleTemplateList, updateRuleTemplateList } = useRuleTemplate();
  const { globalRuleTemplateList, updateGlobalRuleTemplateList } =
    useGlobalRuleTemplate();

  const [projectName, setProjectName] = useState<string>();
  const [ruleTemplateName, setRuleTemplateName] = useState<string>();

  const projectNameChangeHandle = (name: string) => {
    setProjectName(name);
    setRuleTemplateName(undefined);
    if (name) {
      updateRuleTemplateList(name);
    }
  };

  const ruleTemplateNameChangeHandle = (name: string) => {
    setRuleTemplateName(name);
    if (!name) {
      return;
    }
    if (projectName) {
      getProjectTemplateRules(projectName, name);
    } else {
      getGlobalTemplateRules(name);
    }
  };

  const generateRuleTemplateSelectOptions = () => {
    const list = projectName ? ruleTemplateList : globalRuleTemplateList;
    const groupLabel = projectName
      ? t('rule.projectRuleTemplate')
      : t('rule.globalRuleTemplate');

    if (list.length === 0) {
      return null;
    }

    return (
      <Select.OptGroup label={groupLabel}>
        {list.map((v) => {
          return (
            <Select.Option
              key={v.rule_template_name}
              value={v.rule_template_name ?? ''}
            >
              {v.rule_template_name}
            </Select.Option>
          );
        })}
      </Select.OptGroup>
    );
  };

  useEffect(() => {
    updateGlobalRuleTemplateList();
    updateDriverNameList();
    updateProjectList();
  }, [updateDriverNameList, updateGlobalRuleTemplateList, updateProjectList]);

  useEffect(() => {
    if (driverNameList.length > 0 && !dbType) {
      setDbType(driverNameList[0]);
    }
  }, [driverNameList, dbType]);

  useEffect(() => {
    const searchParams = new URLSearchParams(location.search);

    const projectNameInUrl = searchParams.get(RuleUrlParamKey.projectName);
    const ruleTemplateNameInUrl = searchParams.get(
      RuleUrlParamKey.ruleTemplateName
    );
    if (projectNameInUrl) {
      setProjectName(projectNameInUrl);
      updateRuleTemplateList(projectNameInUrl);
    }

    if (ruleTemplateNameInUrl) {
      setRuleTemplateName(ruleTemplateNameInUrl);
      projectNameInUrl
        ? getProjectTemplateRules(projectNameInUrl, ruleTemplateNameInUrl)
        : getGlobalTemplateRules(ruleTemplateNameInUrl);
    }
  }, [
    getGlobalTemplateRules,
    getProjectTemplateRules,
    location.search,
    updateRuleTemplateList,
  ]);

  return {
    generateDriverSelectOptions,
    generateProjectSelectOption,
    generateRuleTemplateSelectOptions,
    projectName,
    projectNameChangeHandle,
    dbType,
    setDbType,
    ruleTemplateName,
    ruleTemplateNameChangeHandle,
  };
};

export default useRuleFilterForm;
