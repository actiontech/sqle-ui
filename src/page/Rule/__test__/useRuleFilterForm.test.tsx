import { waitFor } from '@testing-library/react';
import { act, renderHook } from '@testing-library/react-hooks/dom';
import {
  mockDriver,
  mockUseGlobalRuleTemplate,
  mockUseProject,
  mockUseRuleTemplate,
} from '../../../testUtils/mockRequest';
import useRuleFilterForm from '../useRuleFilterForm';
import route from 'react-router';

describe('test useRuleFilterForm', () => {
  const projectName = 'default';
  const ruleTemplateName = 'rule1';
  let getDriverSpy: jest.SpyInstance;
  let getProjectSpy: jest.SpyInstance;
  let getRuleTemplateSpy: jest.SpyInstance;
  let getGlobalRuleTemplateSpy: jest.SpyInstance;
  const getProjectTemplateRulesSpy = jest.fn();
  const getGlobalTemplateRulesSpy = jest.fn();

  beforeEach(() => {
    getDriverSpy = mockDriver();
    getProjectSpy = mockUseProject();
    getRuleTemplateSpy = mockUseRuleTemplate();
    getGlobalRuleTemplateSpy = mockUseGlobalRuleTemplate();

    jest.useFakeTimers();
    jest.spyOn(route, 'useLocation').mockReturnValue({
      pathname: '/rule',
      hash: '',
      search: '',
      state: '',
    });
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  test('should send request get tips', async () => {
    expect(getDriverSpy).toBeCalledTimes(0);
    expect(getProjectSpy).toBeCalledTimes(0);
    expect(getRuleTemplateSpy).toBeCalledTimes(0);
    expect(getGlobalRuleTemplateSpy).toBeCalledTimes(0);
    renderHook(() =>
      useRuleFilterForm(getProjectTemplateRulesSpy, getGlobalTemplateRulesSpy)
    );

    expect(getDriverSpy).toBeCalledTimes(1);
    expect(getProjectSpy).toBeCalledTimes(1);
    expect(getRuleTemplateSpy).toBeCalledTimes(0);
    expect(getGlobalRuleTemplateSpy).toBeCalledTimes(1);
  });

  test('should set dbType default value when driverNameList length  is greater than zero', async () => {
    const { result } = renderHook(() =>
      useRuleFilterForm(getProjectTemplateRulesSpy, getGlobalTemplateRulesSpy)
    );

    expect(result.current.dbType).toBeUndefined();

    await waitFor(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(result.current.dbType).toBe('oracle');
  });

  test('should be match snapshot of rule template options', async () => {
    const { result } = renderHook(() =>
      useRuleFilterForm(getProjectTemplateRulesSpy, getGlobalTemplateRulesSpy)
    );

    expect(result.current.generateRuleTemplateSelectOptions()).toBeNull();

    await waitFor(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(
      result.current.generateRuleTemplateSelectOptions()
    ).toMatchSnapshot();

    act(() => {
      result.current.projectNameChangeHandle(projectName);
    });
    expect(result.current.projectName).toBe(projectName);
    expect(getRuleTemplateSpy).toBeCalledTimes(1);
    expect(getRuleTemplateSpy).toBeCalledWith({
      project_name: projectName,
    });

    await waitFor(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(
      result.current.generateRuleTemplateSelectOptions()
    ).toMatchSnapshot();

    act(() => {
      result.current.projectNameChangeHandle('');
    });
    expect(getRuleTemplateSpy).toBeCalledTimes(1);
  });

  test('should clear rule template name when changed project name or changed dbType', async () => {
    const { result } = renderHook(() =>
      useRuleFilterForm(getProjectTemplateRulesSpy, getGlobalTemplateRulesSpy)
    );

    await waitFor(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(getGlobalTemplateRulesSpy).toBeCalledTimes(0);

    act(() => {
      result.current.ruleTemplateNameChangeHandle(ruleTemplateName);
    });
    expect(result.current.ruleTemplateName).toBe(ruleTemplateName);

    expect(getProjectTemplateRulesSpy).toBeCalledTimes(0);
    expect(getGlobalTemplateRulesSpy).toBeCalledTimes(1);
    expect(getGlobalTemplateRulesSpy).toBeCalledWith(ruleTemplateName);

    act(() => {
      result.current.projectNameChangeHandle(projectName);
    });

    expect(result.current.ruleTemplateName).toBeUndefined();

    act(() => {
      result.current.ruleTemplateNameChangeHandle(ruleTemplateName);
    });
    expect(getGlobalTemplateRulesSpy).toBeCalledTimes(1);
    expect(getProjectTemplateRulesSpy).toBeCalledTimes(1);
    expect(getProjectTemplateRulesSpy).toBeCalledWith(
      projectName,
      ruleTemplateName
    );

    act(() => {
      result.current.ruleTemplateNameChangeHandle('');
    });
    expect(getGlobalTemplateRulesSpy).toBeCalledTimes(1);
    expect(getProjectTemplateRulesSpy).toBeCalledTimes(1);
  });

  test('should be executed change event with url search parameter', async () => {
    jest.spyOn(route, 'useLocation').mockReturnValue({
      pathname: '/rule',
      hash: '',
      search: `?projectName=${projectName}&ruleTemplateName=${ruleTemplateName}`,
      state: '',
    });

    const { result, rerender } = renderHook(() =>
      useRuleFilterForm(getProjectTemplateRulesSpy, getGlobalTemplateRulesSpy)
    );

    expect(result.current.projectName).toBe(projectName);
    expect(getRuleTemplateSpy).toBeCalledTimes(1);
    expect(getRuleTemplateSpy).toBeCalledWith({
      project_name: projectName,
    });

    expect(getGlobalTemplateRulesSpy).toBeCalledTimes(0);
    expect(getProjectTemplateRulesSpy).toBeCalledTimes(1);
    expect(getProjectTemplateRulesSpy).toBeCalledWith(
      projectName,
      ruleTemplateName
    );

    jest.clearAllMocks();
    jest.clearAllTimers();

    jest.spyOn(route, 'useLocation').mockReturnValue({
      pathname: '/rule',
      hash: '',
      search: `ruleTemplateName=${ruleTemplateName}`,
      state: '',
    });

    rerender();
    expect(getRuleTemplateSpy).toBeCalledTimes(0);
    expect(getProjectTemplateRulesSpy).toBeCalledTimes(0);
    expect(getGlobalTemplateRulesSpy).toBeCalledTimes(1);
    expect(getGlobalTemplateRulesSpy).toBeCalledWith(ruleTemplateName);
  });
});
