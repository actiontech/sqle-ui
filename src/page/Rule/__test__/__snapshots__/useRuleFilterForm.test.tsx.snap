// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test useRuleFilterForm should be match snapshot of rule template options 1`] = `
<OptGroup
  label="rule.globalRuleTemplate"
>
  <Option
    value="global_rule_template_name1"
  >
    global_rule_template_name1
  </Option>
  <Option
    value="global_rule_template_name2"
  >
    global_rule_template_name2
  </Option>
</OptGroup>
`;

exports[`test useRuleFilterForm should be match snapshot of rule template options 2`] = `
<OptGroup
  label="rule.projectRuleTemplate"
>
  <Option
    value="rule_template_name1"
  >
    rule_template_name1
  </Option>
  <Option
    value="rule_template_name2"
  >
    rule_template_name2
  </Option>
</OptGroup>
`;
